### PR TITLE
Trace Support

### PR DESCRIPTION
**Pull request by tt-asaigal**
fyi @tt-aho  tracking all our trace commits here.

### CODE REVIEW

**Review by tt-aho — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Comment by tt-asaigal** [comment]
Passing Post Commit: https://github.com/tenstorrent/tt-metal/actions/runs/9118316143

**Comment by tt-rkim** [comment]
@tt-asaigal What about nightly , big man?
I didnt' check nightly history for you, let me know if you ran

**Review by arakhmati — APPROVED** [review]

File: tt_eager/tensor/tensor.cpp [hunk 1]
```diff
@@ -934,10 +932,10 @@ Tensor allocate_tensor_on_device(const Shape& shape, DataType data_type, Layout
         worker->push_work(
             [shape, data_type, layout, worker, memory_config, device_tensor, worker_index] () mutable {
                 if (memory_config.is_sharded()) {
-                    auto local_tensor = create_sharded_device_tensor(shape, data_type, layout, worker, memory_config);
+                    auto local_tensor = create_device_tensor(shape.value(), data_type, layout, worker, memory_config);
```
Comment (abhullar-tt) [inline tt_eager/tensor/tensor.cpp:935]: the if (memory_config.is_sharded()) conditional seems redundant here

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 2]
```diff
@@ -590,14 +590,20 @@ class SystemMemoryManager {
         }
     }
 
-    void fetch_queue_write(uint32_t command_size_B, const uint8_t cq_id) {
+    void fetch_queue_write(uint32_t command_size_B, const uint8_t cq_id, bool stall_prefetcher = false) {
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:593]: can you add a comment about relationship of setting stall_prefetcher and executing a trace

File: tt_metal/host_api.hpp [hunk 3]
```diff
@@ -465,68 +465,72 @@ void Finish(CommandQueue& cq);
...
  *
  * | Argument     | Description                                                            | Type                          | Valid Range                        | Required |
  * |--------------|------------------------------------------------------------------------|-------------------------------|------------------------------------|----------|
- * | trace        | Trace in which to end the capture                                      | Trace &                       |                                    | Yes      |
+ * | device       | The device holding being traced.                                       | Device *                      |                                    | Yes      |
+ * | cq_id        | The command queue id associated with the trace.                        | uint8_t                       |                                    | Yes      |
+ * | tid          | A unique id from BeginTraceCapture for the trace being captured        | uint32_t                      |                                    | Yes      |
  */
-void EndTrace(Trace &trace);
+void EndTraceCapture(Device *device, const uint8_t cq_id, const uint32_t tid);
```
Comment (abhullar-tt) [inline tt_metal/host_api.hpp:492]: Can you add a note about disabling/enabling allocations in the docs?

**Review by abhullar-tt — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

**Review by tapspatel — APPROVED** [review]

**Review by pgkeller — APPROVED** [review]

**Pull request merged**